### PR TITLE
Name and description definition in headers

### PR DESCRIPTION
**Pull request by SafarMirek**
This PR implements feature requested in #1629.
User can provide name or description in headers of new artifact or new version upload request.

`X-Registry-Name` or `X-Registry-Name-Encoded` - artifact name
`X-Registry-Description` or `X-Registry-Description-Encoded` - artifact description

If user provides both non-encoded and encoded values HTTP Conflict is raised.

### CODE REVIEW

File: app/src/main/java/io/apicurio/registry/rest/v2/GroupsResourceImpl.java [hunk 1]
```diff
@@ -469,14 +480,22 @@ public void deleteArtifactsInGroup(String groupId) {
...
-            IfExists ifExists, Boolean canonical, InputStream data) {
+    @Authorized(style = AuthorizedStyle.GroupOnly, level = AuthorizedLevel.Write)
+    public ArtifactMetaData createArtifact(String groupId, ArtifactType xRegistryArtifactType, String xRegistryArtifactId,
+                                           String xRegistryVersion, IfExists ifExists, Boolean canonical,
+                                           String xRegistryDescription, String xRegistryDescriptionEncoded,
+                                           String xRegistryName, String xRegistryNameEncoded, InputStream data) {
         requireParameter("groupId", groupId);
+
+        maxOneOf("xRegistryName", xRegistryName, "xRegistryNameEncoded", xRegistryNameEncoded);
+        maxOneOf("xRegistryDescription", xRegistryDescription, "xRegistryDescriptionEncoded", xRegistryDescriptionEncoded);
```
Comment (EricWittmann) [inline app/src/main/java/io/apicurio/registry/rest/v2/GroupsResourceImpl.java:494]: I think the names you pass in here should be the actual HTTP header names:
* X-Registry-Name
* X-Registry-Name-Encoded
* X-Registry-Description
* X-Registry-Description-Encoded

Same for all of the calls to `maxOneOf` elsewhere.

File: app/src/main/java/io/apicurio/registry/rest/v2/GroupsResourceImpl.java [hunk 2]
```diff
@@ -594,6 +624,23 @@ private static final void requireParameter(String parameterName, Object paramete
         }
     }
 
+    private static void maxOneOf(String parameterOneName, Object parameterOneValue, String parameterTwoName, Object parameterTwoValue) {
+        if (parameterOneValue != null && parameterTwoValue != null) {
+            throw new ConflictException("Both " + parameterOneName + " and " + parameterTwoName + " are present.");
```
Comment (EricWittmann) [inline app/src/main/java/io/apicurio/registry/rest/v2/GroupsResourceImpl.java:629]: Let's change the language here to:  "Conflict: '" + parameterOneName + "' and '" + parameterTwoName + "' are mutually exclusive."

File: app/src/main/java/io/apicurio/registry/storage/dto/PartlyFilledArtifactMetaDataDto.java [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+
+package io.apicurio.registry.storage.dto;
+
+import io.quarkus.runtime.annotations.RegisterForReflection;
+
+import java.util.List;
+import java.util.Map;
+
+@RegisterForReflection
+public class PartlyFilledArtifactMetaDataDto extends EditableArtifactMetaDataDto {
```
Comment (EricWittmann) [inline app/src/main/java/io/apicurio/registry/storage/dto/PartlyFilledArtifactMetaDataDto.java:26]: I don't think we need this class.  I think you can just use the `EditableArtifactMetaDataDto` class directly, no?

File: app/src/main/java/io/apicurio/registry/storage/impl/sql/AbstractSqlRegistryStorage.java [hunk 4]
```diff
@@ -2719,15 +2720,18 @@ protected ContentHandle canonicalizeContent(ArtifactType artifactType, ContentHa
         }
     }
 
-    protected EditableArtifactMetaDataDto extractMetaData(ArtifactType artifactType, ContentHandle content) {
+    protected EditableArtifactMetaDataDto extractMetaData(ArtifactType artifactType, ContentHandle content, EditableArtifactMetaDataDto metaData) {
+        if (metaData == null) {
+            metaData = new EditableArtifactMetaDataDto();
+        }
```
Comment (EricWittmann) [inline app/src/main/java/io/apicurio/registry/storage/impl/sql/AbstractSqlRegistryStorage.java:2726]: It's not usually a good idea to do this, although it is convenient.  In this case I don't think we want to pass in the metaData to this method at all, so I would ask that you revert this change and we can keep calling this method only if no meta-data was provided by the user.

File: app/src/test/java/io/apicurio/registry/rest/v2/GroupsResourceTest.java [hunk 5]
```diff
@@ -295,6 +297,87 @@ public void testCreateArtifact() throws Exception {
...
+                .post("/registry/v2/groups/{groupId}/artifacts")
+            .then()
+                .statusCode(200)
+                .body("groupId", equalTo(GROUP))
+                .body("description", equalTo(customDescription))
+                .body("id", equalTo("testCreateArtifact/EmptyAPI-customDescription"))
+                .body("type", equalTo(ArtifactType.OPENAPI.name()));
+
+        // Create OpenAPI artifact - provide a custom No-ASCII description
+        String customNoASCIIDescription = "CUSTOM DESCRIPTION with NO-ASCII char č";
```
Comment (EricWittmann) [inline app/src/test/java/io/apicurio/registry/rest/v2/GroupsResourceTest.java:352]: This will be interesting.  How does our toolchain handle unicode characters?  :)

**Comment by EricWittmann** [comment]
Good start, and nice job on the tests!  I think I would prefer to keep the previous logic where we only extract meta-data from the content if no custom meta-data was provided.  So I suggest we keep that as the contract, which I think would allow you to remove the new class you introduced, along with the instanceof checks.  Instead, let's only create an `EditableArtifactMetaData` instance if meta-data was actually provided.  This is instead of always creating one and later checking if it has any non-null values.  Just need a convenience method to create an instance or return null depending on whether a name or description is provided.

Thoughts everyone?

**Comment by SafarMirek** [comment]
So, if I have understood it right, it should work this way:
If user had defined name or description in headers we would not extract any other parameter from the content. If so, it would make it much easier. 
In this version, It don't extract only values that user had specified. This is reason why it pass the metadata object into extraction method.

File: app/src/main/java/io/apicurio/registry/rest/v2/GroupsResourceImpl.java (hunk 2 above)
Comment (carlesarnal) [reply app/src/main/java/io/apicurio/registry/rest/v2/GroupsResourceImpl.java:629]: This exception should not be used here. This ConflictException is being used in the compatibility layer and is subject to change there and should not be used in the rest layer. You can use one of the existing ones or you think this is worth a new one just create it.

File: app/src/main/java/io/apicurio/registry/storage/dto/PartlyFilledArtifactMetaDataDto.java (hunk 3 above)
Comment (carlesarnal) [reply app/src/main/java/io/apicurio/registry/storage/dto/PartlyFilledArtifactMetaDataDto.java:26]: +1

File: app/src/main/java/io/apicurio/registry/storage/impl/sql/AbstractSqlRegistryStorage.java (hunk 4 above)
Comment (carlesarnal) [reply app/src/main/java/io/apicurio/registry/storage/impl/sql/AbstractSqlRegistryStorage.java:2726]: +1

File: common/src/main/resources/META-INF/openapi.json [hunk 6]
```diff
@@ -0,0 +1,3270 @@
+{
```
Comment (carlesarnal) [inline common/src/main/resources/META-INF/openapi.json:1]: I guess something bad happened here when adding the new changes right? :)

**Comment by carlesarnal** [comment]
> Good start, and nice job on the tests! I think I would prefer to keep the previous logic where we only extract meta-data from the content if no custom meta-data was provided. So I suggest we keep that as the contract, which I think would allow you to remove the new class you introduced, along with the instanceof checks. Instead, let's only create an `EditableArtifactMetaData` instance if meta-data was actually provided. This is instead of always creating one and later checking if it has any non-null values. Just need a convenience method to create an instance or return null depending on whether a name or description is provided.
> 
> Thoughts everyone?

I agree with that approach. The biggest problem I see right now with this is the use of the ConflictException from the compatibility layer but that's a very easy fix.

**Comment by SafarMirek** [comment]
I have implemented your suggestions. Thanks

**Comment by EricWittmann** [comment]
Looks good, @SafarMirek !  Just click the "Ready for review" button and I think we can merge this asap.

@carlesarnal should we also update the REST client to support the new headers?

**Comment by carlesarnal** [comment]
> Looks good, @SafarMirek ! Just click the "Ready for review" button and I think we can merge this asap.
> 
> @carlesarnal should we also update the REST client to support the new headers?

Good catch. We don't strictly need to change anything in the REST client to support this but we might want to add an additional method to the REST client with these values to have something more user-friendly. Right now, users must be aware of this feature and use `[setNextRequestHeaders](https://github.com/Apicurio/apicurio-registry/blob/master/rest-client/client-common/src/main/java/io/apicurio/registry/rest/client/RegistryClient.java#L171)` in their REST client to send these values as headers.

@SafarMirek let me know if you want to add that method to the REST client, we can have a quick call and I can guide you to the required changes there. Otherwise, I'll probably do it asap.
